### ISSUE TITLE
[clang-tidy] NOLINT not working for cert-err58-cpp inside a macro

### ISSUE THREAD

**Issue by fekir** [llvm/llvm-project#191963]
Here is a minimal example:

```c++
#include <string>

#define CREATE_STRUCT(name) struct name {   /* NOLINT */ \
    static inline const std::string b = "1378462394867239567252385623563483464836386"; \
}

CREATE_STRUCT(x);
```

Then execute `clang-tidy test.cpp -checks=cert-err58-cpp`.

I get the following diagnostic

```
Running without flags.
2 warnings generated.
warning: initialization of 'b' with static storage duration may throw an exception that cannot be caught [cert-err58-cpp]
    9 | CREATE_STRUCT(X);
      | ^
note: expanded from macro 'CREATE_STRUCT'
    5 |     static inline const std::string b = "1378462394867239567252385623563483464836386"; \
      |                                     ^
```

`NOLINTBEGIN`, `NOLINTEND` and `NOLINTNEXTLINE` also do not work, using `NOLINT(cert-err58-cpp)` instead of `NOLINT` does not change anything.

Tested with

```
LLVM (http://llvm.org/):
  LLVM version 22.1.3
  Optimized build.
```

Possibly related bug https://github.com/llvm/llvm-project/issues/55134 and https://github.com/llvm/llvm-project/issues/46511, although the provided test-cases do not trigger a warning, thus it seems specific to `cert-err58-cpp`, and not a general issue.

**Comment by IamYJLee** [comment]
I’d like to try fixing this issue.

**Comment by IamYJLee** [comment]
@zeyi2 
I think we should change `NoLintDirectiveHandler`.
When I changed the code below in ThrowingStaticInitializationCheck.cpp, it only removed the warning note.
https://github.com/llvm/llvm-project/blob/72a3cd9e54f16d87f47bc984672999a9b67cd4f4/clang-tools-extra/clang-tidy/bugprone/ThrowingStaticInitializationCheck.cpp#L42-L45

## Changed
```cpp
SourceLocation Loc = VD->getLocation();
  if (Loc.isMacroID())
    Loc = Result.Context->getSourceManager().getExpansionLoc(Loc);

  diag(Loc,
       "initialization of %0 with %select{static|thread_local}1 storage "
       "duration may throw an exception that cannot be caught")
      << VD << (VD->getStorageDuration() == SD_Static ? 0 : 1);
```

## Result 
### Original
```
/Users/yjlee/Documents/dev/llvm-project/build-debug/test.cpp:7:1: warning: initialization of 'b' with static storage duration may throw an exception that cannot be caught [cert-err58-cpp]
    7 | CREATE_STRUCT(x);
      | ^
/Users/yjlee/Documents/dev/llvm-project/build-debug/test.cpp:4:37: note: expanded from macro 'CREATE_STRUCT'
    4 |     static inline const std::string b = "1378462394867239567252385623563483464836386"; \
      |                                     ^
```
### Modified
```
/Users/yjlee/Documents/dev/llvm-project/build-debug/test.cpp:7:1: warning: initialization of 'b' with static storage duration may throw an exception that cannot be caught [cert-err58-cpp]
    7 | CREATE_STRUCT(x);
      | ^
```

Could you advise me on how I can solve this issue?

**Comment by IamYJLee** [comment]
And both of the following cases work without any code changes.
```cpp
#include <string>
#define CREATE_STRUCT(name) struct name {  /* NOLINTNEXTLINE */ \
    static inline const std::string b = "1378462394867239567252385623563483464836386";   \
}
CREATE_STRUCT(x);
```
```cpp
#include <string>
#define CREATE_STRUCT(name) struct name { \
    static inline const std::string b = "1378462394867239567252385623563483464836386";   /* NOLINT */  \
}
CREATE_STRUCT(x);
```

**Comment by vbvictor** [comment]
> And both of the following cases work without any code changes.
> 
> #include <string>
> #define CREATE_STRUCT(name) struct name {  /* NOLINTNEXTLINE */ \
>     static inline const std::string b = "1378462394867239567252385623563483464836386";   \
> }
> CREATE_STRUCT(x);
> #include <string>
> #define CREATE_STRUCT(name) struct name { \
>     static inline const std::string b = "1378462394867239567252385623563483464836386";   /* NOLINT */  \
> }
> CREATE_STRUCT(x);

I checked this (https://godbolt.org/z/sq8b1fv8o) and don't see any bug.
Diagnostic is emitted for `b` variable (not struct name) and `nolint` should be on the line of diagnostic. Everything work as expected

**Comment by IamYJLee** [comment]
Yes, I checked it on the main branch as well, and it looks fine.

**Comment by vbvictor** [comment]
> Yes, I checked it on the main branch as well, and it looks fine.

FYI, godbolt usually has 1-2 day old main branch (probably rebuilt daily) so It's pretty accurate in representing current state of things

**Comment by IamYJLee** [comment]
@fekir @zeyi2 @vbvictor 
Can I close this issue?
